arangod/Aql/OptimizerRulesCluster.cpp

```diff
@@ -0,0 +1,450 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2018 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Wilfried Goesgens
+/// @author Jan Christoph Uhde
+////////////////////////////////////////////////////////////////////////////////
+
+#include "OptimizerRules.h"
+#include "Aql/ClusterNodes.h"
+
+#include "Aql/ExecutionNode.h"
+#include "Aql/IndexNode.h"
+#include "Aql/ModificationNodes.h"
+
+#include "Aql/ExecutionPlan.h"
+#include "Aql/Condition.h"
+
+#include "Aql/Optimizer.h"
+
+#include <boost/optional.hpp>
+#include <tuple>
+
+using namespace arangodb;
+using namespace arangodb::aql;
+using EN = arangodb::aql::ExecutionNode;
+
+// in plan below
+ExecutionNode* hasSingleParent(ExecutionNode* in){
```

Comment:
function is still a (less efficient) copy of `ExecutionNode::getFirstParent()`!

Reply:
How is it a copy if one function checks for empty and the other for the size of one? Maybe front() is slower with -O0 I bet (1 Käsebrötchen) that front and the subscription operator will behave the same with  -O3.